{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "vcf,bed, with or without gender"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "python liftOverPlink.py --map myplink.map --out lifted --chain hg19ToHg38.over.chain.gz\n",
    "plink --vcf update3.vcf.gz --maf 0.05 --recode --out myplink\n",
    "plink --vcf update3.vcf.gz --biallelic-only strict --make-bed --impute-sex ycount --out tz_test1\n",
    "plink --vcf update3.vcf.gz --biallelic-only strict --make-bed --out tz_test\n",
    "plink --bfile tz_plink_ld_r2_0.5 --pca 20 --out tz_plink_ld_r2_0.5\n",
    "plink --bfile bin_final_strand --extract ld_r2_0.5.prune.in --make-bed --out tz_plink_ld_r2_0.5\n",
    "plink --bfile bin_final_strand --extract ld_r2_0.2.prune.in --make-bed --out tz_plink_ld_r2_0.2\n",
    "plink --bfile bin_final_strand --indep-pairphase 20000 2000 0.2 --make-bed --out ld_r2_0.2\n",
    "\n",
    "plink --vcf update3.vcf.gz --maf 0.05 --recode --out myplink\n",
    "python liftOverPlink.py --map myplink.map --out lifted --chain hg38ToHg19.over.chain.gz\n",
    "\n",
    "python liftOverPlink.py --map myplink.map --out lifted --chain hg19ToHg38.over.chain.gz\n",
    "\n",
    "plink --bfile tz_test1 --recode --out tz_test1\n",
    "python liftOverPlink.py --map tz_test1.map --out lifted --chain hg38ToHg19.over.chain.gz\n",
    "liftOver tz_test1.lift hg38ToHg19.over.chain.gz tz_test1.newfile tz_test1.unmaped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'height_pipeline' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-8ec227aa9086>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mheight_pipeline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'ddat/crowdAI/temp'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mchain\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'hg38ToHg19'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'linear_regression'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msnp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'top200_all'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'full_model'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'height_pipeline' is not defined"
     ]
    }
   ],
   "source": [
    "height_pipeline('ddat/crowdAI/temp',chain='hg38ToHg19',ms=['linear_regression'],snp='top200_all',path='full_model')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def init_args(arguments=None):\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('-g', '--genotype', help='Genotype input file, the prefix of plink files (.bim,.fam,.bed)')\n",
    "    parser.add_argument('-c', '--chain', help='Liftover Chain file: hg38ToHg19 or hg19ToHg38',default='hg38ToHg19')\n",
    "    parser.add_argument('--gender', help='Specify gender for non-label specimen')\n",
    "    parser.add_argument('-l', '--genotype-label', help='Whether has genotype label', action='store_true')\n",
    "    parser.add_argument('--no-fill', help='Do not fill NA genotypes', action='store_true')\n",
    "    parser.add_argument('--no-drop', help='Do not drop rows or columns', action='store_true')\n",
    "    parser.add_argument('--drop-row-thresh', help='Drop row if number of NA greater than this threshold', type=int, default=30)\n",
    "    parser.add_argument('--drop-col-thresh', help='Drop column if number of NA greater than this threshold', type=int, default=50)\n",
    "    parser.add_argument('-v', '--verbose', help='Show more information', action='store_true')\n",
    "    parser.add_argument('--snp-list', help='Choose type of snp list', choices=['1', '5', '10', '15', '25', '43', '100', '119', '1996'], default='119')\n",
    "    parser.add_argument('-e', '--source', help='Input file source', choices=['none', '23andme', '23andme-exome-vcf', 'ftdna-illumina', 'decodeme', 'vcf', 'ftdna'], default='none')\n",
    "    parser.add_argument('--min-snp', help='Min SNP number for non-labeled mode.', type=int)\n",
    "    args = parser.parse_args(arguments)\n",
    "    if not args.genotype or not os.path.exists(args.genotype):\n",
    "        print('Invalid genotype file')\n",
    "        return None\n",
    "    if not args.output:\n",
    "        print('Invalid output file')\n",
    "        return None\n",
    "    return args\n",
    "\n",
    "def main(args=None):\n",
    "    args = init_args(args)\n",
    "    liftover_plink(args.genotype,args.chain)\n",
    "\n",
    "        \n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_file='ddat/crowdAI/update3.vcf.gz'\n",
    "def height_pipeline(g,meta=None,chain='hg38ToHg19',ms=['linear_regression'],snp='top200_all',path='full_model'):\n",
    "    plink_path = parse_genotype(g)\n",
    "    print('plink_path',plink_path)\n",
    "    lift_file = liftover_plink(plink_path,chain)\n",
    "    print('liftover_plink',lift_file)\n",
    "    height_prediction(plink_path,lift_file,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:27<00:00, 12.32s/it]\n",
      "/home/ahmadA/dereck/proj/predict_height/snp2h.py:61: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  lift_bim['pos'] = list(lift.end)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "snp match ratio : 0.9833948339483395\n",
      "higher is better. if the ration is less than 0.8, the result is not reliable\n",
      "    chrom        snp_x  cm_x        pos a0_x a1_x  i_x        snp_y  cm_y  \\\n",
      "0       1     rs600933   0.0    4409496    C    T    0     rs600933   0.0   \n",
      "1       1    rs6684802   0.0   16449680    T    C    1    rs6684802   0.0   \n",
      "2       1    rs9435734   0.0   17311190    G    A    2    rs9435734   0.0   \n",
      "3       1   rs11261067   0.0   18455424    A    G    3   rs11261067   0.0   \n",
      "4       1   rs16861635   0.0   18471296    C    T    4   rs16861635   0.0   \n",
      "5       1  kgp11946142   0.0   31904584    T    C    5    rs4949405   0.0   \n",
      "6       1      rs14528   0.0   40219280    C    T    6      rs14528   0.0   \n",
      "7       1     rs914962   0.0   40242315    T    C    7     rs914962   0.0   \n",
      "8       1    rs3131683   0.0   40513710    C    T    8    rs3131683   0.0   \n",
      "9       1   kgp2140909   0.0   40935087    T    C    9   rs72950069   0.0   \n",
      "10      1   rs17110383   0.0   54790499    T    C   10   rs17110383   0.0   \n",
      "11      1    rs2819483   0.0   56160151    T    C   11    rs2819483   0.0   \n",
      "12      1   kgp9447679   0.0   56783362    G    T   12   rs12030407   0.0   \n",
      "13      1    rs3005906   0.0   56790383    C    T   13    rs3005906   0.0   \n",
      "14      1    rs3005911   0.0   56794965    G    A   14    rs3005911   0.0   \n",
      "15      1    rs2811881   0.0   59228062    T    C   15    rs2811881   0.0   \n",
      "16      1   rs17127673   0.0   65955725    G    A   16   rs17127673   0.0   \n",
      "17      1   rs12086324   0.0   69681190    T    C   17   rs12086324   0.0   \n",
      "18      1     rs617211   0.0   70990438    G    T   18     rs617211   0.0   \n",
      "19      1  kgp15637460   0.0   75448817    G    A   19   rs77290506   0.0   \n",
      "20      1   kgp3928043   0.0   90302825    G    A   20   rs67951971   0.0   \n",
      "21      1   rs11807256   0.0  100770760    G    T   21   rs11807256   0.0   \n",
      "22      1   rs11166440   0.0  100808363    A    G   22   rs11166440   0.0   \n",
      "23      1   kgp3704429   0.0  100989434    A    G   23   rs75076995   0.0   \n",
      "24      1    rs4329498   0.0  119133609    C    T   24    rs4329498   0.0   \n",
      "25      1    rs1631505   0.0  119179191    C    T   25          NaN   NaN   \n",
      "26      1   rs16836032   0.0  154450730    C    A   26   rs16836032   0.0   \n",
      "27      1   rs16832061   0.0  160461060    T    C   27   rs16832061   0.0   \n",
      "28      1   rs10494532   0.0  180690420    T    G   28   rs10494532   0.0   \n",
      "29      1   rs10914083   0.0  180700133    A    G   29   rs10914083   0.0   \n",
      "..    ...          ...   ...        ...  ...  ...  ...          ...   ...   \n",
      "512    21    rs9976372   0.0   41760095    C    T  512    rs9976372   0.0   \n",
      "513    21    rs9974272   0.0   42521711    G    A  513    rs9974272   0.0   \n",
      "514    21    rs9984948   0.0   43127316    A    C  514    rs9984948   0.0   \n",
      "515    22    rs1807460   0.0   23505934    T    C  515    rs1807460   0.0   \n",
      "516    22   kgp2260059   0.0   31058193    G    A  516    rs9606761   0.0   \n",
      "517    22    rs2014661   0.0   45865055    T    C  517    rs2014661   0.0   \n",
      "518    23    rs5935702   0.0   14119191    C    T  518    rs5935702   0.0   \n",
      "519    23    rs5969772   0.0   16914295    A    G  519    rs5969772   0.0   \n",
      "520    23    rs5969770   0.0   16917522    A    C  520    rs5969770   0.0   \n",
      "521    23  kgp22832049   0.0   16922536    G    T  521    rs1000657   0.0   \n",
      "522    23    rs6527831   0.0   18105267    A    C  522    rs6527831   0.0   \n",
      "523    23   rs10521917   0.0   24012381    C    T  523   rs10521917   0.0   \n",
      "524    23   rs12840034   0.0   32106875    T    C  524   rs12840034   0.0   \n",
      "525    23    rs6632511   0.0   36206175    A    G  525    rs6632511   0.0   \n",
      "526    23    rs4419977   0.0   36218964    T    C  526    rs4419977   0.0   \n",
      "527    23    rs6632558   0.0   36315793    C    T  527    rs6632558   0.0   \n",
      "528    23    rs6629072   0.0   36318935    G    A  528    rs6629072   0.0   \n",
      "529    23   rs12011068   0.0   67376320    T    C  529   rs12011068   0.0   \n",
      "530    23    rs7061649   0.0   71128234    T    C  530    rs7061649   0.0   \n",
      "531    23    rs7878231   0.0   71178847    A    G  531    rs7878231   0.0   \n",
      "532    23    rs4986610   0.0   71209899    G    A  532    rs4986610   0.0   \n",
      "533    23   rs12011214   0.0   71210482    T    C  533   rs12011214   0.0   \n",
      "534    23  kgp22746254   0.0   71219960    A    G  534   rs12006999   0.0   \n",
      "535    23  kgp22780824   0.0   75987774    T    C  535  rs147291751   0.0   \n",
      "536    23    rs6621367   0.0   79220877    A    G  536    rs6621367   0.0   \n",
      "537    23  kgp22803904   0.0   79231118    C    T  537    rs1429595   0.0   \n",
      "538    23  kgp22795119   0.0   79231201    T    C  538    rs6523640   0.0   \n",
      "539    23  kgp22817007   0.0  113314976    C    T  539  rs149816616   0.0   \n",
      "540    23  kgp22760083   0.0    2022542    T    C  540    rs6644924   0.0   \n",
      "541    23  kgp22803856   0.0    2535604    G    A  541    rs5939245   0.0   \n",
      "\n",
      "    a0_y a1_y         i_y  \n",
      "0      T    C     27549.0  \n",
      "1      T    C    104903.0  \n",
      "2      G    A    108286.0  \n",
      "3      A    G    116616.0  \n",
      "4      C    T    116744.0  \n",
      "5      T    C    193482.0  \n",
      "6      C    T    237019.0  \n",
      "7      T    C    237096.0  \n",
      "8      C    T    238859.0  \n",
      "9      T    C    241190.0  \n",
      "10     T    C    312092.0  \n",
      "11     T    C    321719.0  \n",
      "12     G    T    325339.0  \n",
      "13     C    T    325394.0  \n",
      "14     G    A    325435.0  \n",
      "15     T    C    341188.0  \n",
      "16     G    A    379014.0  \n",
      "17     T    C    399999.0  \n",
      "18     T    G    406691.0  \n",
      "19     G    A    429892.0  \n",
      "20     G    A    516672.0  \n",
      "21     G    T    570963.0  \n",
      "22     A    G    571168.0  \n",
      "23     A    G    572037.0  \n",
      "24     C    T    678292.0  \n",
      "25   NaN  NaN         NaN  \n",
      "26     C    A    722850.0  \n",
      "27     T    C    756564.0  \n",
      "28     T    G    874177.0  \n",
      "29     A    G    874231.0  \n",
      "..   ...  ...         ...  \n",
      "512    C    T  16170364.0  \n",
      "513    G    A  16176018.0  \n",
      "514    A    C  16180026.0  \n",
      "515    T    C  16254747.0  \n",
      "516    G    A  16302785.0  \n",
      "517    T    C  16395972.0  \n",
      "518    C    T  16520113.0  \n",
      "519    A    G  16531209.0  \n",
      "520    A    C  16531223.0  \n",
      "521    G    T  16531247.0  \n",
      "522    A    C  16534832.0  \n",
      "523    C    T  16559696.0  \n",
      "524    T    C  16593790.0  \n",
      "525    A    G  16615043.0  \n",
      "526    T    C  16615089.0  \n",
      "527    C    T  16615529.0  \n",
      "528    G    A  16615547.0  \n",
      "529    T    C  16718695.0  \n",
      "530    T    C  16732891.0  \n",
      "531    A    G  16733052.0  \n",
      "532    G    A  16733176.0  \n",
      "533    T    C  16733178.0  \n",
      "534    A    G  16733198.0  \n",
      "535    T    C  16752169.0  \n",
      "536    A    G  16766518.0  \n",
      "537    C    T  16766626.0  \n",
      "538    T    C  16766627.0  \n",
      "539    C    T  16916565.0  \n",
      "540    T    C  16461253.0  \n",
      "541    G    A  16465305.0  \n",
      "\n",
      "[542 rows x 12 columns]\n",
      "the ratio of genotype shift 0.04127579737335835\n",
      "Success genotype shift\n",
      "              linear_regression\n",
      "chenshiting          159.522489\n",
      "gengguangyao         161.385321\n",
      "leichen              163.042709\n",
      "leijunqin            164.578576\n",
      "luopin               162.421226\n",
      "qiaolu               160.284737\n",
      "qinpengfei           169.732352\n",
      "tangxiangjie         168.400195\n",
      "L00044               167.221817\n",
      "L00045               168.347535\n",
      "L00046               165.184147\n",
      "L00047               161.428551\n",
      "L00048               161.842466\n",
      "L00049               169.509545\n",
      "L00050               153.464714\n",
      "L00051               157.703710\n",
      "L00052               158.937839\n",
      "L00053               160.466771\n",
      "L00054               165.259973\n",
      "L00055               151.234573\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ahmadA/dereck/proj/predict_height/snp2h.py:62: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  lift_bim['i'] = list(lift.score)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>linear_regression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chenshiting</th>\n",
       "      <td>159.522489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gengguangyao</th>\n",
       "      <td>161.385321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>leichen</th>\n",
       "      <td>163.042709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>leijunqin</th>\n",
       "      <td>164.578576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>luopin</th>\n",
       "      <td>162.421226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>qiaolu</th>\n",
       "      <td>160.284737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>qinpengfei</th>\n",
       "      <td>169.732352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tangxiangjie</th>\n",
       "      <td>168.400195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00044</th>\n",
       "      <td>167.221817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00045</th>\n",
       "      <td>168.347535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00046</th>\n",
       "      <td>165.184147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00047</th>\n",
       "      <td>161.428551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00048</th>\n",
       "      <td>161.842466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00049</th>\n",
       "      <td>169.509545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00050</th>\n",
       "      <td>153.464714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00051</th>\n",
       "      <td>157.703710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00052</th>\n",
       "      <td>158.937839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00053</th>\n",
       "      <td>160.466771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00054</th>\n",
       "      <td>165.259973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L00055</th>\n",
       "      <td>151.234573</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              linear_regression\n",
       "chenshiting          159.522489\n",
       "gengguangyao         161.385321\n",
       "leichen              163.042709\n",
       "leijunqin            164.578576\n",
       "luopin               162.421226\n",
       "qiaolu               160.284737\n",
       "qinpengfei           169.732352\n",
       "tangxiangjie         168.400195\n",
       "L00044               167.221817\n",
       "L00045               168.347535\n",
       "L00046               165.184147\n",
       "L00047               161.428551\n",
       "L00048               161.842466\n",
       "L00049               169.509545\n",
       "L00050               153.464714\n",
       "L00051               157.703710\n",
       "L00052               158.937839\n",
       "L00053               160.466771\n",
       "L00054               165.259973\n",
       "L00055               151.234573"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "height_prediction('ddat/crowdAI/temp','ddat/crowdAI/temp.Hg19',{'ms':['linear_regression'],'snp':'top200_all','path':'full_model'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas_plink import read_plink\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from LiftoverCmd import liftover_plink\n",
    "\n",
    "from PlinkCmd import vcf2plink,ped2plink,guess_gender\n",
    "\n",
    "from snp2h import height_feature,height_predict\n",
    "from predh import read_bim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def init_args(arguments=None):\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument('-g', '--genotype', help='Genotype input file, the prefix of plink files (.bim,.fam,.bed)')\n",
    "    parser.add_argument('-s', '--specimen', help='the sample information')\n",
    "    parser.add_argument('-l', 'liftover', help='the output file with suffix .Hg19 or .Hg38 from LiftoverCmd.py',default=None)\n",
    "    #parser.add_argument('-c', '--chain', help='Liftover Chain file',choices=['hg38ToHg19','hg19ToHg3'],default='hg38ToHg19')\n",
    "    parser.add_argument('-m', '--model', help='a list of prediction models',default=['linear_regression'])\n",
    "    parser.add_argument('--snp-list', help='Choose type of snp list', choices=['top200','top200_all','top300','top300_all','top500'], default='top200_all')\n",
    "    parser.add_argument('--model-path',help='Choose the way of model generation',choices=['full_model','rep10cv10_model'], default= 'full_model')\n",
    "\n",
    "    \n",
    "    parser.add_argument('--gender', help='Specify gender for non-label specimen')\n",
    "    parser.add_argument('-l', '--genotype-label', help='Whether has genotype label', action='store_true')\n",
    "    parser.add_argument('--no-fill', help='Do not fill NA genotypes', action='store_true')\n",
    "    parser.add_argument('--no-drop', help='Do not drop rows or columns', action='store_true')\n",
    "    parser.add_argument('--drop-row-thresh', help='Drop row if number of NA greater than this threshold', type=int, default=30)\n",
    "    parser.add_argument('--drop-col-thresh', help='Drop column if number of NA greater than this threshold', type=int, default=50)\n",
    "    parser.add_argument('-v', '--verbose', help='Show more information', action='store_true')\n",
    "    parser.add_argument('--snp-list', help='Choose type of snp list', choices=['1', '5', '10', '15', '25', '43', '100', '119', '1996'], default='119')\n",
    "    parser.add_argument('-e', '--source', help='Input file source', choices=['none', '23andme', '23andme-exome-vcf', 'ftdna-illumina', 'decodeme', 'vcf', 'ftdna'], default='none')\n",
    "    parser.add_argument('--min-snp', help='Min SNP number for non-labeled mode.', type=int)\n",
    "    args = parser.parse_args(arguments)\n",
    "    if not args.genotype or not os.path.exists(args.genotype):\n",
    "        print('Invalid genotype file')\n",
    "        return None\n",
    "    if not args.output:\n",
    "        print('Invalid output file')\n",
    "        return None\n",
    "    return args\n",
    "\n",
    "def main(args=None):\n",
    "    args = init_args(args)\n",
    "    height_prediction(args.genotype,args.liftover,args.specimen,ms=args.model,snp=args.snp_list,path=args.model_path)\n",
    "        \n",
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
